Allow Klippy to listen to a TCP socket

Toniob opened the pull request:
This PR is to allow Klippy to listen to a TCP socket. The changes are backward compatible with the current configuration. It allows to use unix:// url to bind to a unix socket or tcp:// for the new features.

I’ve modified whconsole.py to test my patch and everything seems to be fine.

JamesH1978 commented:
Thankyou for submitting a PR, please be aware you need to sign off, as per point 3 in https://github.com/Klipper3d/klipper/blob/master/docs/CONTRIBUTING.md#what-to-expect-in-a-review

Thanks
James

Toniob committed:
klippy: Allow to listen to a TCP socket
docs: add instructions to use TCP sockets

This changes the way that the apiserver parameter is handled. It now
uses a URL instead of a string which was the path of the unix socket.
The string representing a path is still allowed, if no scheme is present
it considers its a unix socket. I also log errors if klippy cannot bind
to a file or the IP, instead of raising an exception.

Signed-off-by: Anthony Bourguignon <contact@toniob.net>

Toniob commented:
Sorry for this oversight. I’ve just changed my commit. I hope it’s better this way.

lexfrei commented:
@Toniob I can suggest to update the docs too

Toniob committed:
docs: add instructions on how to listen to a TCP socket

Change the API server doc to add informations on the new url to bind on
a unix socket or a tcp socket.

Signed-off-by: Anthony Bourguignon <contact@toniob.net>

KevinOConnor commented:
Thanks.  From the conversation at https://klipper.discourse.group/t/make-the-api-listens-to-a-tcp-socket/21939/13 it seems there is some concern from some of the other long-term contributors (eg, security, duplication with moonraker, limited audience).  I'd be inclined to hold off on this until there's a clear audience of a notable size that requires something like this (we typically aim for a minimum of at least a 100 users for changes to the main Klipper repo): https://www.klipper3d.org/CONTRIBUTING.html

-Kevin

lexfrei commented:
@KevinOConnor look at https://hub.docker.com/r/mkuf/moonraker please. We can't speak about users count here, but 50k downloads is a big number for a non-official image. We can divide 50k by tags count (120) assuming that each user used each tag once and it's 400+ of users. Each of them needs this feature.

lisbonc1987 commented:
Very interested in this. Some printers have resource issues and this would enable to free up resources.

alund1990 commented:
Need this now.

NetBUG commented:
Cool, great fix/addition

shvetsm commented:
I would love to see this added, this will be a boon for people with cheapo printers

ankushg commented:
Another user reporting that I'd love this feature

Would also happily accept some amount of limitations to mitigate the very valid security concerns if it moves us in the direction of getting this merged into klipper and moonraker - maybe requiring a shared 256-bit secret between the two?

jatinkrmalik commented:
I would also love this one!

farhankamili commented:
I'd also like this feature to be implemented.
